crates/starknet-devnet-server/src/api/json_rpc/endpoints_ws.rs

```diff
@@ -0,0 +1,111 @@
+
+/// The definitions of JSON-RPC read endpoints defined in starknet_ws_api.json
+impl JsonRpcHandler {
+    pub async fn execute_ws(
+        &self,
+        request: JsonRpcSubscriptionRequest,
+        rpc_request_id: Id,
+        socket_id: SocketId,
+    ) -> Result<(), ApiError> {
+        match request {
+            JsonRpcSubscriptionRequest::NewHeads(data) => {
+                self.subscribe_new_heads(data, rpc_request_id, socket_id).await
+            }
+            JsonRpcSubscriptionRequest::TransactionStatus => todo!(),
+            JsonRpcSubscriptionRequest::PendingTransactions => todo!(),
+            JsonRpcSubscriptionRequest::Events => todo!(),
+            JsonRpcSubscriptionRequest::Unsubscribe(SubscriptionIdInput { subscription_id }) => {
+                let mut sockets = self.api.sockets.lock().await;
+                let socket_context = sockets.get_mut(&socket_id).ok_or(
+                    ApiError::StarknetDevnetError(Error::UnexpectedInternalError {
+                        msg: format!("Unregistered socket ID: {socket_id}"),
+                    }),
+                )?;
+
+                socket_context.unsubscribe(rpc_request_id, subscription_id).await?;
+                Ok(())
+            }
+        }
+    }
+
+    /// starknet_subscribeNewHeads
+    /// Checks if an optional block ID is provided. Validates that the block exists and is not too
+    /// many blocks in the past. If it is a valid block, the user is notified of all blocks from the
+    /// old up to the latest, and subscribed to new ones. If no block ID specified, the user is just
+    /// subscribed to new blocks.
+    pub async fn subscribe_new_heads(
+        &self,
+        block_id_input: Option<BlockIdInput>,
+        rpc_request_id: Id,
+        socket_id: SocketId,
+    ) -> Result<(), ApiError> {
+        let latest_tag = BlockId::Tag(BlockTag::Latest);
+        let block_id = if let Some(BlockIdInput { block_id }) = block_id_input {
+            block_id.into()
+        } else {
+            // if no block ID input, this eventually just subscribes the user to new blocks
+            latest_tag
+        };
+
+        let starknet = self.api.starknet.lock().await;
+
+        // checking the block's existence; aborted blocks treated as not found
+        let query_block = match starknet.get_block(&block_id) {
+            Ok(block) => match block.status() {
+                BlockStatus::Rejected => Err(ApiError::BlockNotFound),
+                _ => Ok(block),
+            },
+            Err(Error::NoBlock) => Err(ApiError::BlockNotFound),
+            Err(other) => Err(ApiError::StarknetDevnetError(other)),
+        }?;
+
+        let latest_block = starknet.get_block(&latest_tag)?;
+
+        let query_block_number = query_block.block_number().0;
+        let latest_block_number = latest_block.block_number().0;
+
+        let blocks_back_amount = if query_block_number > latest_block_number {
```

Comment:
if user cant specify block number in the future, then this if clause is not useful. `query_block_number` will be at least equal to `latest_block_number`

Reply:
Added this check to prevent overflow (unsigned integer subtraction can't produce a negative result)

Reply:
is it possible to have overflow when `query_block_number` can't be greater than `latest_block_number` because of this comment: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/634#discussion_r1832657640

Reply:
It depends on whether we want to define a custom block invalid error now or wait for the spec to be updated.